### PR TITLE
add facebook authorization ability

### PR DESCRIPTION
**Pull request by ashonuyak**
See - https://app.asana.com/0/1202525679663009/1202829131337119/f

### CODE REVIEW

File: backend/src/middlewares/auth/strategies/facebook.strategy.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      'hometown',
+      'location',
+      'birthday',
+      'picture',
+    ],
+    passReqToCallback: true,
+  },
+  async (req, accessToken, refreshToken, profile, done) => {
+    console.log(profile);
+    console.log(parseInt(profile._json.birthday.substr(-4)));
```
Comment (oleksandr-danylchenko) [inline backend/src/middlewares/auth/strategies/facebook.strategy.ts]: Remove logs of the sensitive data

File: backend/src/middlewares/auth/strategies/facebook.strategy.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    let user;
```
Comment (oleksandr-danylchenko) [inline backend/src/middlewares/auth/strategies/facebook.strategy.ts]: Why did you declare the untyped `user` variable here?  Couldn't it be declared inside the try-catch?

File: backend/src/middlewares/auth/strategies/facebook.strategy.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        where: {
+          User_Security: {
+            facebook_acc_id: profile.id,
+          },
+        },
+      });
+
+      if (user) return done(null, user);
+
+      const payload: string | JwtPayload | undefined =
```
Comment (oleksandr-danylchenko) [inline backend/src/middlewares/auth/strategies/facebook.strategy.ts]: Why did you need to specify types of the `payload` explicitly here?

File: frontend/src/components/sign/components/sign-with-facebook/sign-with-facebook.tsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    return (): void => {
+      window.removeEventListener('storage', onStorageEvent, false);
+    };
+  }, [onStorageEvent]);
+
+  return (
+    <ButtonOutline text={title + ' with Facebook'} onClick={onFacebookSign} />
+  );
+};
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/sign/components/sign-with-facebook/sign-with-facebook.tsx]: This component is almost an identical copy of the Facebook handling page. Could you please make a genetic component that will work both for Google and Facebook?

File: frontend/src/components/sign/components/sign-with-facebook/sign-with-facebook.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+const SignWithFacebook = ({ title }: { title: string }): React.ReactElement => {
+  const navigate = useNavigate();
+  const dispatch = useAppDispatch();
+
+  const onStorageEvent = useCallback((e: StorageEvent): void => {
+    if (e.key === 'token') {
+      const accessToken = localStorage.getItem('token');
+      dispatch(setCredentials({ accessToken }));
+      navigate('/', { replace: true });
```
Comment (oleksandr-danylchenko) [inline frontend/src/components/sign/components/sign-with-facebook/sign-with-facebook.tsx]: Where's the refresh token?

File: frontend/src/helpers/sign/open-facebook.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { ENV } from '@common/enums/app/env.enum';
+
+const openFacebookAuthPage = (token: string | null = null): void => {
+  window.open(
+    `${ENV.API_PATH}/auth/facebook/sign?token=${token}`,
+    'Facebook OAuth',
+    'height=615,width=605',
+  );
+};
```
Comment (oleksandr-danylchenko) [inline frontend/src/helpers/sign/open-facebook.ts]: This function almost duplicates the google auth opener. Extract the reusable function

File: backend/src/routes/auth/auth.router.ts [hunk 7]
```diff
@@ -23,7 +23,15 @@ authRouter.get(`${PATH}/google/sign`, passportMiddleware.googleAuth);
...
-  authController.signGoogle,
+  authController.signGoogleFacebook,
+);
+
+authRouter.get(`${PATH}/facebook/sign`, passportMiddleware.facebookAuth);
+
+authRouter.get(
+  `${PATH}/facebook/redirect`,
+  passportMiddleware.facebookMiddleware,
+  authController.signGoogleFacebook,
```
Comment (oleksandr-danylchenko) [inline backend/src/routes/auth/auth.router.ts]: Don't use the name of the exact 3rd party provider name if the controller isn't done exclusively for it. Because if we would decide to add Discord or Microsoft providers, the name would grow more and more. Use more generic name, like "sighOAuthProviders"

**Review by oleksandr-danylchenko — CHANGES_REQUESTED** [review]
You're going the right direction, but some code style/duplication fixes required

File: frontend/src/components/sign/components/sign-with-facebook/sign-with-facebook.tsx (hunk 5 above)
Comment (ashonuyak) [reply frontend/src/components/sign/components/sign-with-facebook/sign-with-facebook.tsx]: we just dont have refresh token in state

File: frontend/src/components/sign/components/sign-with-oauth/sign-with-oauth.tsx [hunk 8]
```diff
@@ -33,7 +33,7 @@ const SignWithGoogle = ({ title }: { title: string }): React.ReactElement => {
     };
   }, [onStorageEvent]);
 
-  return <ButtonOutline text={title + ' with Google'} onClick={onGoogleSign} />;
+  return <ButtonOutline text={'Sign In with ' + title} onClick={onOAuthSign} />;
```
Comment (R0ixy) [inline frontend/src/components/sign/components/sign-with-oauth/sign-with-oauth.tsx]: It looks like we gonna display _Sign In_ on both `Sign In` and `Sign Up` pages. Maybe it will be better if we change it to `Continue with`

File: frontend/src/components/sign/components/sign-with-facebook/sign-with-facebook.tsx (hunk 5 above)
Comment (oleksandr-danylchenko) [reply frontend/src/components/sign/components/sign-with-facebook/sign-with-facebook.tsx]: We must add the refresh token to the response. It can be done in the next task

File: backend/src/middlewares/auth/strategies/google.strategy.ts [hunk 9]
```diff
@@ -13,9 +13,8 @@ const googleStrategy = new GoogleStrategy(
     scope: ['profile', 'email'],
   },
   async (req, accessToken, refreshToken, profile, done) => {
-    let user;
     try {
-      user = await prisma.user.findFirst({
+      let user = await prisma.user.findFirst({
```
Comment (oleksandr-danylchenko) [inline backend/src/middlewares/auth/strategies/google.strategy.ts:17]: Why not `const`? Will the `user` be reassigned?

**Review by oleksandr-danylchenko — CHANGES_REQUESTED** [review]

File: backend/src/middlewares/auth/strategies/google.strategy.ts (hunk 9 above)
Comment (ashonuyak) [reply backend/src/middlewares/auth/strategies/google.strategy.ts:17]: Yeah, it's reassigned in the code below

Comment (oleksandr-danylchenko) [reply backend/src/middlewares/auth/strategies/google.strategy.ts:17]: Okay, I just saw it below :ok_hand:

**Review by oleksandr-danylchenko — APPROVED** [review]
